com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java

```diff
@@ -0,0 +1,174 @@
+import com.sap.cloud.lm.sl.common.util.CommonUtil;
+import com.sap.cloud.lm.sl.common.util.JsonUtil;
+
+@Component("determineServiceCreateUpdateActionsStep")
+public class DetermineServiceCreateUpdateServiceActionsStep extends SyncFlowableStep {
+
+    @Inject
+    private ServiceInstanceGetter serviceInstanceGetter;
+
+    private SecureSerializationFacade secureSerializer = new SecureSerializationFacade();
+
+    @Override
+    protected StepPhase executeStep(ExecutionWrapper execution) throws Exception {
+        execution.getStepLogger().info("Entering determine state...");
+        CloudControllerClient controllerClient = execution.getControllerClient();
+        String spaceId = StepsUtil.getSpaceId(execution.getContext());
+
+        CloudServiceExtended serviceToProcess = StepsUtil.getServiceToProcess(execution.getContext());
+        CloudService existingService = controllerClient.getService(serviceToProcess.getName(), false);
+
+        Map<String, List<String>> defaultTags = computeDefaultTags(controllerClient);
+        List<String> serviceDefaultTags = defaultTags.getOrDefault(serviceToProcess.getLabel(), Collections.emptyList());
+
+        Map<String, List<ServiceKey>> serviceKeys = StepsUtil.getServiceKeysToCreate(execution.getContext());
+
+        List<ServiceAction> actions = determineActions(controllerClient, spaceId, serviceToProcess, existingService, serviceDefaultTags,
+            serviceKeys, execution);
+        StepsUtil.setServiceActions(actions, execution.getContext());
+
+        return StepPhase.DONE;
+    }
+
+    private List<ServiceAction> determineActions(CloudControllerClient client, String spaceId, CloudServiceExtended service,
+        CloudService existingService, List<String> defaultTags, Map<String, List<ServiceKey>> serviceKeys, ExecutionWrapper execution) {
+        List<ServiceAction> actions = new ArrayList<>();
+
+        getStepLogger().debug("Determining action to be performed on a service...");
+
+        if (existingService == null) {
+            getStepLogger().debug("Service should be created");
+            getStepLogger().debug("New service: " + secureSerializer.toJson(service));
+            return Arrays.asList(ServiceAction.CREATE);
+        }
+
+        // Check if the existing service should be updated or not
+        if (shouldRecreate(service, existingService)) {
+            getStepLogger().debug("Service should be recreated");
+            getStepLogger().debug("New service: " + secureSerializer.toJson(service));
+            getStepLogger().debug("Existing service: " + secureSerializer.toJson(existingService));
+            StepsUtil.setServicesToDelete(execution.getContext(), Arrays.asList(service.getName()));
+            return Arrays.asList(ServiceAction.RECREATE);
+        } else {
+            //Overriding possibly inherited values
+            StepsUtil.setServicesToDelete(execution.getContext(), Collections.EMPTY_LIST);
```

Comment:
Why doing it here? Why setting it at all?

Reply:
Because we do inherit the variables from the parent context, which mens that something can be already populated in this variable.
The logic will be changed to not inherit variables and work only with the required subset of vars.